### ISSUE TITLE
Support for response caching

### ISSUE THREAD

**Issue by jkonecki** [damianh/ProxyKit#36]
I would like to cache responses received from upstream servers based on standard http caching headers.

My scenario involves some slow-changing reference data that could be cached by ProxyKit and save a network trip to the upstream server.

It would be nice to be able to enable caching for individual routes, but in my scenario that isn't necessary.

Happy to contribute when given some guidelines.

**Comment by jkonecki** [comment]
I know Ali (the world is so small ;-) )
That would be cool - the only question is how to handle the dependency.
Separate NuGet package maybe?

**Comment by jkonecki** [comment]
I've done a spike and the caching can be simply implemented using:

```
var client = ClientExtensions.CreateClient();
app.RunProxy(context =>
{
    var forwardContext = context.ForwardTo("upstream-host-url");

    return client.SendAsync(forwardContext.UpstreamRequest);
});
```

Since CacheCow supports using `HttpRequestMessage` and `HttpResponseMessage`, it integrates nicely with `Execute()`.

I would live to live it to you to decide on the best integration design.

**Comment by damianh** [comment]
Good start. The small problem with this is that is it not using the HttpClient factory.

**Comment by damianh** [comment]
> Separate NuGet package maybe?

I _think_ I'll go along with recipe right now as opposed to maintaining integration packages.

**Comment by jkonecki** [comment]
I assume, Damian, that you would like CacheCow to use the HttpClient instance injected into ForwardingContext instead of using own instance?

**Comment by damianh** [comment]
No, that's dead code that I forgot to remove (I'll remove it now). The code of interest is:

https://github.com/damianh/ProxyKit/blob/3b1f34d072b13e60ed639d1f7edf6b5c9c14160f/src/ProxyKit/ProxyContextExtensions.cs#L31-L33

**Comment by jkonecki** [comment]
Does it mean that if I register HttpClient with CacheCow caching handler
than it should be picked up here and used by ProxyKit?
Or am I missing something?

**Comment by damianh** [comment]
`SharedProxyOptions` has a place to insert a custom HttpMessageHandler that is used when creating the HttpMessageHandler pipeline when creating the HttpClient that is injected into `ProxyKitClient` (aka ["typed http client"](https://www.hanselman.com/blog/HttpClientFactoryForTypedHttpClientInstancesInASPNETCore21.aspx)). Am going to see if it's easy to inject `CachingHandler` here. 

The way HttpClientFactory works is by disposing and re-creating all the handlers on a timed interval. I have a funny feeling that ProxyKit's API as it currently is might not quite be as simple as desired.

**Comment by jkonecki** [comment]
Hooking up a static instance of
https://github.com/aliostad/CacheCow/blob/master/src/CacheCow.Client/CachingHandler.cs
should do the job, shouldn't it?

**Comment by damianh** [comment]
Can't be static. HttpClientFactory will want to dispose it and recreate it. This can result in `ObjectDisposedException`s a couple of minutes into a running system (which doesn't show up in tests). Related https://github.com/damianh/ProxyKit/pull/8

**Comment by jkonecki** [comment]
Ok. There is an overload that takes a Cache store. I can create a single
static instance of the store but keep recreating the handler itself.

**Comment by damianh** [comment]
Yes, singleton CacheStore

**Comment by damianh** [comment]
So this appears to work now:

```csharp
var cacheStore = new InMemoryCacheStore(TimeSpan.FromMinutes(1));
services.AddProxy(options =>
{
	options.GetMessageHandler = () => new CachingHandler(cacheStore)
	{
		InnerHandler = new HttpClientHandler {AllowAutoRedirect = false, UseCookies = false}
	};
});
```

Not quite pit-of-success...

1. `options.GetMessageHandler` is hooked into the HttpClient Factory as the Primary handler. But here CachingHandler is a `DelegatingHandler` and I didn't put hook inplace for that. Hence manually setting the InnderHandler..
2. We're not leveraging the IoC container to activate `InMemoryCacheStore` nor `CachingHandler` (with ICacheStore being injected into it).

Need to think about this API a bit.

Example CacheCow test code here: https://github.com/damianh/ProxyKit/blob/cachecow/src/ProxyKit.Tests/CacheCowTests.cs#L19

**Comment by jkonecki** [comment]
Just came up with the same myself ;-)

I don't really mind about not using DI for cache store - the whole proxy
thanks to your lib fits on one screen :-D

**Comment by jkonecki** [comment]
I think it may be safer to hold own instance of InMemoryVaryHeaderStore as
well

**Comment by damianh** [comment]
You do need to strongly manage the lifecycles of those, yes.

**Comment by jkonecki** [comment]
Thank you very much, Damian!
All working lovely in production :-)

**Comment by damianh** [comment]
Production already? Lovely! 😄 

Based on this discussion, I made some improvements to make this easier: #37 . This change was semver breaking so new major version.

Also added recipe for CacheCow and am supporting it via tests

 - https://github.com/damianh/ProxyKit/blob/master/src/Recipes/11_CachingWithCacheCow.cs
 - https://github.com/damianh/ProxyKit/blob/master/src/ProxyKit.Tests/CacheCowTests.cs

I don't think an integration package is worth it right now but will keep open mind to it if there is more demand and makes things easier.

Thanks for suggestion and the help!

**Comment by jkonecki** [comment]
I think the recipe is a perfect solution - allows for full customisation
which would require a lot of work if CacheCow was hidden behind ProxyKit
façade.
